Scanner

pmengelbert opened the pull request:
**What this PR does / why we need it**:
- Add new `scanner` binary that uses trivy to scan for vulnerabilities in images.
- Enable scanner to read from Collector CRs, and write to Collector CR status.
- Provide make targets to build scanner images.


**Special notes for your reviewer**:
- In the interest of moving things along, this PR uses the REST client and the manager's `serviceaccount`. I opened #177 to track this

pmengelbert committed:
Add initial scanner implementation

* Add trivy v0.24.4 as a dependency
* Add initial scanner implementation, which currently doesn't work

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Fix scanner code to work with trivy v0.24.4

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Add makefile target for scanner images; Add to dockerfile for scanner

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Add ImageCollector permissions

Via magic comments in imagejob_controller.go

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Work on REST client approach

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Scanner will now update status of collector cr

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Get rid of hardcoded strings

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

sozercan commented on pkg/scanner/scanner.go:
```diff
@@ -0,0 +1,345 @@
+package main
```
should this go under something like /pkg/scanners/trivy/trivy.go?

sozercan commented on pkg/scanner/scanner.go:
```diff
@@ -0,0 +1,345 @@
+
+	sopts := trivyTypes.ScanOptions{
+		VulnType:            []string{"os", "library"},
+		SecurityChecks:      []string{"vuln", "secret"},
```
vulnType and securityChecks should be configurable with flags

sozercan commented on pkg/scanner/scanner.go:
```diff
@@ -0,0 +1,345 @@
+	for _, sev := range sevs {
+		_, ok := severityMap[sev]
+		if !ok {
+			return fmt.Errorf("severity '%s' should be one of of [CRITICAL, HIGH, MEDIUM, LOW, UNKNOWN]", sev)
```
```suggestion
			return fmt.Errorf("severity '%s' should be one of [CRITICAL, HIGH, MEDIUM, LOW, UNKNOWN]", sev)
```

sozercan commented on Dockerfile:
```diff
@@ -30,6 +31,16 @@ RUN \
 
+FROM builder AS scanner-build
+
+ARG TARGETOS
```
can we define ARG at the top? do these need to be duplicated?

pmengelbert committed:
Remove duplicated ARGs in Dockerfile

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Fix typo

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

sozercan commented on Dockerfile:
```diff
@@ -3,6 +3,7 @@ ARG BUILDERIMAGE="golang:1.17"
 
 ARG ERASERBASEIMAGE="gcr.io/distroless/static:latest"
 ARG MANAGERBASEIMAGE="gcr.io/distroless/static:nonroot"
+ARG SCANNERBASEIMAGE="gcr.io/distroless/static:nonroot"
```
instead of generic "scanner", let's say "TRIVY_SCANNER" or something

sozercan replied:
applies to everywhere we say "scanner"

sozercan commented on pkg/scanner/scanner.go:
```diff
@@ -0,0 +1,345 @@
+	collectorCRName = flag.String("collector-cr-name", "collector-cr", "name of the collector cr to read from and write to")
+	cacheDir        = flag.String("cache-dir", "/var/lib/trivy", "path to the cache dir")
+	severity        = flag.String("severity", "CRITICAL,HIGH,MEDIUM,LOW,UNKNOWN", "list of severity levels to report")
+	ignoreUnfixed   = flag.Bool("ignore-unfixed", false, "report only fixed vulnerabilities")
```
should we default to true?

pmengelbert replied:
👍

sozercan commented on pkg/scanner/scanner.go:
```diff
@@ -0,0 +1,345 @@
+var (
+	collectorCRName = flag.String("collector-cr-name", "collector-cr", "name of the collector cr to read from and write to")
+	cacheDir        = flag.String("cache-dir", "/var/lib/trivy", "path to the cache dir")
+	severity        = flag.String("severity", "CRITICAL,HIGH,MEDIUM,LOW,UNKNOWN", "list of severity levels to report")
```
should we default to CRITICAL?

sozercan replied:
thoughts @miwithro?

pmengelbert replied:
Trivy default is `CRITICAL,HIGH`, I'm fine with either that or `CRITICAL`

miwithro replied:
@sozercan I think if we hit MEDIUM, HIGH, CRITICAL we will be able to deliver on customer expectations.

pmengelbert committed:
Allow for configuration of vuln types and security checks

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Fix whitespace mess

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Change flag defaults

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Reorganize scanner package

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
rename scanner -> trivy-scanner

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Tidying to please the linter

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
check if longer timeout helps

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Fix conflicting change

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Reconcile rebase casualties

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

sozercan commented on pkg/scanners/trivy/trivy.go:
```diff
@@ -0,0 +1,399 @@
+package main
```
as we discussed before, let's add unit tests

we can add e2e once we have the collector logic

pmengelbert committed:
Move flag parsing to main function

The `init` function is called during testing, and extra flags get passed
by `go test` that cause a problem.

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Add unit tests for trivvy scanner functions

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Use TARGETPLATFORM for trivy-scanner output stage

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

sozercan commented on pkg/scanners/trivy/trivy.go:
```diff
@@ -0,0 +1,397 @@
+
+	vulnerableImages := make([]eraserv1alpha1.Image, 0, len(scanList))
+	for imageRef := range imgChan {
+		image := eraserv1alpha1.Image{Digest: "abc123", Name: imageRef}
```
placeholder?

sozercan commented on pkg/scanners/trivy/trivy.go:
```diff
@@ -0,0 +1,397 @@
+		close(imgChan)
+	}()
+
+	vulnerableImages := make([]eraserv1alpha1.Image, 0, len(scanList))
```
should we separate out the reporting back logic so it can be reused later?

pmengelbert replied:
@sozercan Would you mind elaborating? I'm not sure I follow what you mean

sozercan replied:
each provider processes collector results and outputs a list, right? say, if we have a different scanner tomorrow, can we share logic for the output of a list?

sozercan commented on pkg/scanners/trivy/trivy.go:
```diff
@@ -0,0 +1,397 @@
+func downloadDB(cacheDir string) error {
+	client := dlDb.NewClient(cacheDir, true)
+	ctx := context.Background()
+	needsUpdate, err := client.NeedsUpdate("dev", false)
```
what's `dev` here?

pmengelbert replied:
It refers to the trivy version. This made its way in from looking at Trivy's state when it was running. I've corrected this now.

pmengelbert committed:
Use Mutex to lock DB during access

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Report failed scan attempts to ImageCollector CR

Scans can fail for several reasons, including lack of registry
authentication, images not existing, and images without names. In order
to maintain transparency, it is useful to report to the CR when the
scanner is unable to process an image.

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Tidy code

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Rename struct for clarity

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Improve trivy scanner's logging

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Remove hardcoded 'dev' string in scanner config

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Improve log message

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

sozercan commented on pkg/scanners/trivy/trivy.go:
```diff
@@ -0,0 +1,435 @@
+		vulnTypeLibrary: false,
+	}
+
+	log = logf.Log.WithName("eraser")
```
do we want to say "scanner" here? and perhaps `WithValues("provider", "trivy")`?

sozercan commented on pkg/scanners/trivy/trivy.go:
```diff
@@ -0,0 +1,435 @@
+				return
+			}
+
+			fmt.Printf("scanning: %s\n", imageRef)
```
should we use log here?

sozercan commented on pkg/scanners/trivy/trivy.go:
```diff
@@ -0,0 +1,435 @@
+	apiPath         = "apis/eraser.sh/v1alpha1"
+	resourceName    = "imagecollectors"
+	subResourceName = "status"
+	trivyVersion    = "v0.24.4"
```
i am not familiar with this, does this need to match the go.mod version of trivy?

pmengelbert replied:
No, not necessarily, but it probably should. Could we set this using LDFlags based on the version in `go.mod`?

sozercan replied:
Looks like trivy gets this info from appVersion, can we use that or is that part of cli only? https://github.com/aquasecurity/trivy/blob/1b66b77f691f2249a63c284056afcad704c0e15f/pkg/commands/option/global.go#L35

pmengelbert replied:
@sozercan I think it's CLI only. I am setting this using ldflags in the makefile:
```
TRIVY_VERSION ?= $(shell go list -f '{{ .Version }}' -m github.com/aquasecurity/trivy)
...
TRIVY_SCANNER_LDFLAGS ?= $(LDFLAGS)-X 'main.trivyVersion=$(TRIVY_VERSION)'
```

pmengelbert replied:
@sozercan This is only used for error output. Trivy's [process](https://github.com/aquasecurity/trivy/blob/06659f1509a19f601d7f7ff0cfe221f87c466566/pkg/db/db.go#L79) for fetching db looks like this:
- Look at the db file on disk, if the db file doesn't exist yet, then this is the first run and it will automatically fetch the latest version of the db, using an OCI client, from `ghcr`. It uses major version tags to fetch the latest version
- the `trivy` library depends on the `trivy-db` package, and that package has a constant `db.Version` ([here](https://github.com/aquasecurity/trivy-db/blob/39d8a6798e072884afa497c4939036dbafd3b2e4/pkg/db/db.go#L20)) set to the major version of the db
- Internally, if the `db.Version` is compared against the `SchemaVersion` found in the db on the filesystem. If `db.Version` is less than `SchemaVersion`, it means we are using an older client with a newer db, and `trivyVersion` will be propagated into the error output.
- If, `db.Version` is >= `SchemaVersion`, then `db.Version` (the version determined by the `trivy-db` version, which is determined by the `trivy` version) will be used as the major version tag for the OCI fetch.
- Finally, timestamps are compared to determine whether or not the db needs to be updated or not.

sozercan commented on pkg/scanners/trivy/trivy.go:
```diff
@@ -0,0 +1,435 @@
+		}
+
+		log.Info(logMsg, "img", img)
+		vulnerableImages = append(vulnerableImages, img.Image)
```
are we adding images that failed to scan to the vulnerable images list?

pmengelbert replied:
@sozercan We are adding them to the `Results` section of `ImageCollector.Status`, with `ScanSucceeded` set to `false`

pmengelbert committed:
Update scanner build after rebase

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Update logging

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Restore args to build stages in Dockerfile

These are actually not propagated unless we explicitly add them to each
build stage.

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Set trivy version using ldflags for scanner

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Update LDFLAGS in Makefile

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

sozercan reviewed (approved):
Thank you! LGTM
